File: LpSolve/Elements/HalfSpace.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using LpSolve.Interface;

namespace LpSolve.Elements
{
	public class HalfSpace : IElement<HalfSpace>
	{
		private Plane _plane;

		public Plane Plane { get { return this._plane; } }

		public HalfSpace(Plane plane, bool isOnPositive)
		{
			this._plane = plane;

			if (!isOnPositive)
			{
				this._plane.Vector.Flip();
			}
		}

		public bool Contains(Point p)
		{
			var vect = Vector.CreateFromPoints(this._plane.Point, p);

			return vect.ScalarProduct(this._plane.Vector) >= 0;
		}

		public HalfSpace MoveDown(Plane plane)
		{
			var pl = this._plane.MoveDown(plane);

			if (pl == null)
			{
				return null;
			}

			return new HalfSpace(pl, true);
		}

		public int GetDimension()
		{
			return this._plane.GetDimension();
		}
	}
}

File: cpp/Elements/HalfSpace.cpp
#include "HalfSpace.h"
#include "Plane.h"
#include "Point.h"
#include "Vector.h"

using namespace LpSolveCpp::Interface;
namespace LpSolveCpp
{
	namespace Elements
	{

		LpSolveCpp::Elements::Plane *HalfSpace::getPlane() const
		{
			return this->_plane;
		}

		HalfSpace::HalfSpace(LpSolveCpp::Elements::Plane *plane, bool isOnPositive)
		{
			this->_plane = plane;

			if (!isOnPositive)
			{
				this->_plane->getVector()->Flip();
			}
		}

		bool HalfSpace::Contains(Point *p)
		{
			auto vect = Vector::CreateFromPoints(this->_plane->getPoint(), p);

			return vect->ScalarProduct(this->_plane->getVector()) >= 0;
		}

		HalfSpace *HalfSpace::MoveDown(LpSolveCpp::Elements::Plane *plane)
		{
			auto pl = this->_plane->MoveDown(plane);

			if (pl == nullptr)
			{
				return nullptr;
			}

			return new HalfSpace(pl, true);
		}

		int HalfSpace::GetDimension()
		{
			return this->_plane->GetDimension();
		}
	}
}
